Improve footer validation

robertgzr opened the pull request:
relates to #36

robertgzr committed:
test: Add test to showcase #36

Signed-off-by: Robert Günzler <robertg@balena.io>

robertgzr committed:
Validate footers with the grammar

We want to only allow specified footer tags to be recognized as valid.
Otherwise the grammar accepts anything similar to `<tag>: <value>`
It's still possible to allow arbitrary tags via the configuration.

Signed-off-by: Robert Günzler <robertg@balena.io>

stathismor committed:
Allow body tags

Allow tags in body and correctly get the footers when they exist

Change-type: patch
Signed-off-by: Stathis Moraitidis <stathis@balena.io>

stathismor commented on lib/grammar/grammar.pegjs:
```diff
@@ -70,14 +68,22 @@ Body "body"
-	= $([A-Za-z\-]+ ': ' notEOL+) / ( '(cherry picked from commit 'i [0-9a-f]+ ')' {})
+Footer "footer" =
+	('(cherry picked from commit ' notEOL+ {}) /
+	$(t:([a-zA-Z\-]+ Colon) & {
```
What is `t:`'s usage here?

robertgzr replied:
it allows the `& {}` rule to access the matched text

stathismor commented on lib/grammar/grammar.pegjs:
```diff
@@ -70,14 +68,22 @@ Body "body"
+Footer "footer" =
+	('(cherry picked from commit ' notEOL+ {}) /
+	$(t:([a-zA-Z\-]+ Colon) & {
+		return config['valid-trailers'].some((valid) => {
```
Is it here where we should make this check case-insensitive (allow `Change-type` and `change-type`) ?

robertgzr replied:
hmm depends on how we want to use the valid-trailers config field
you could write:
```
{
	"valid-trailers": [
		"[Cc]hange-type",
	]
}
```
to accept both

stathismor replied:
Up to you, but I think all titled/non-titled permutations are allowed. Just saw you change, I don't know about the rest, but `Change-Type` is definitely also allowed. So maybe allow all words to be case insensitive. Like `[Cc]hange-[Tt]ype` and so on.

robertgzr replied:
or always make the regex match case-insensitive?

robertgzr replied:
that would make the config field less powerful and require the `pretty-tags` rule, which we could technically drop after this lands...

stathismor reviewed (commented):
Isn't the terminology we use footer and footer tags, rather than trailers?

robertgzr committed:
rules: check that valid footers don't appear in the body

Signed-off-by: Robert Günzler <robertg@balena.io>

robertgzr committed:
valid-footer-tags: validate against the entire footer

this allows explicitely matching values as well

Signed-off-by: Robert Günzler <robertg@balena.io>

robertgzr commented:
* renamed all uses of the word trailer to footer
* removed the new test case and added the change to the `well-formed-commit` test case instead
* added a check that ensures valid footers don't appear in the body (previous `no-tags-in-body` behaviour)
* changed the grammar to validate the full footer line (instead of only the tag before the `:`)

this allows:
```
{
	"valid-footer-tags": [
		"^Change-type: (major|minor|patch)$"
	]
}
```

robertgzr committed:
valid-footer-tags: Add additional valid tags

Signed-off-by: Robert Günzler <robertg@balena.io>

robertgzr commented:
includes #50
